### ISSUE TITLE
Question about the calculation of the derivation for the loss

### ISSUE THREAD

**Issue by oceanypt** [hughperkins/rationalizing-neural-predictions#1]
I have the problem to understand the calculation of the derivation for the loss in Tao's paper. Your codes support my original understanding. But could you please explain why we calculate the loss function like this(shown in your code)?

**Comment by hughperkins** [comment]
There are three parts to the loss:
- encouraging the model to correctly predict the target sentiments, this is the `loss_mse`, which is standard MSE loss
- encourage the number of selected words to be small. This is `loss_z1`, which is L1 loss, and is proportional to the number of selected words. So, fewer selected words => smaller loss, and
- encourage the selected words to be adjacent to each other, this is `loss_transition`, which is the number of contiguous word selections

I guess your question relates mostly to the `loss_transition`?  The idea is, say we have the passage:

```
the beer was n’t what i expected, and i‘m not sure it’s “true to style“, but i thought it was delicious. a very pleasant ruby red-amber color with a relatively brilliant finish, but a limited amount of carbonation, from the look of it. aroma is what i think an amber ale should be - a nice blend of caramel and happiness bound together.
```
We'd rather the model chose a contiguous set of words, like say:
```
pleasant ruby red-amber color
```
... than choose words scattered around, even if the number of selected words is the same, eg:
```
pleasant red-amber brilliant happiness
```

To what extent does this somewhat address your question? Which parts of the loss are you looking into most?

**Comment by oceanypt** [comment]
Hi,

Actually, I am confused about the sampling technic to train generator, mostly related to loss_mse.

What is sampling? For my opinion, I think this is about: to train the model, in the feedforward process, we will use the untrained gen to select the words whose z is equal to 1, when the loss is backward, we will not train the words whose z is equal to 0. Is it right?

Thanks,
Hai Ye

**Comment by hughperkins** [comment]
loss_mse will encourage both the generator and encoder to learn:
- the encoder only sees the words selected by the generator, and will receive therefore a different input sample on each forward propagation, each time that the generator forwards a different sentence to it
- so the encoder will, as you say, only learn based on the set of words the generator passes to it, for each example. you are exactly right
- thinking about this, this might be a bit confusing to the encoder, since it might see a sentence like 'awesome aroma' and a color rating of '0 stars', and then learns on this occasion to associate 'awesome' and 'aroma' with '0 stars'
- hopefully, this will average out over many examples though, and empirically, it does

Thoughts? To what extent does this explanation sound plausible-ish?

**Comment by yangshao** [comment]
Hi, hughperkins
     I also have a similar question on the loss calculation. In tao's paper, the loss update has two parts one's for generator, and another for encoder. I think the related codes are:

loss = loss_mse + sparsity * loss_z1 + coherence * loss_transitions
rationale_selected_node.reinforce(-loss.data[0])
loss.backward(rationale_selected_node)

I think the reinforce line code is used to update the parameter of the generator. But I don't quite get the meaning of "loss.backward(rationale_selected_node)". why do we need to pass the parameter rationale_selected_node? 

Thanks.
